File: experiments/export/export_train.rb

require "rubygems"
require "active_record"

class DB < ActiveRecord::Base
  set_table_name "template_mo"
end

class ExportTrain
  
  def initialize
    @fields = ["store_key","title_key","on_year","on_month", "sales_total"]
  end
  
  def execute_query
    rs =  DB.find_by_sql("select #{@fields.join(',')} from template_mo")
    raise "Query failed" if rs.nil? or rs.size<=0
    return rs
  end
  
  def load
    puts "Perforing query..."
    rs = execute_query
    puts "Got #{rs.size} records"
    lines = Array.new(rs.size)
    rs.each_with_index do |row, i|
      values = []
      @fields.each {|f| values << row[f]}
      lines[i] = values.join(',')
      puts "processed #{i} lines..." if (i%20000)==0 
    end
    write_output_file(lines, "export_train.dat")
  end


  def write_output_file(lines_array, filename)
    f = File.open(filename, 'w')
    lines_array.each {|line| f.write(line + "\n")} 
    f.close
    puts "Wrote #{lines_array.length} lines to file '#{filename}'"
  end
  
end

if __FILE__ == $0
  ActiveRecord::Base.establish_connection(:adapter=>"mysql", :host => "localhost",
    :username=>"root", :database=>"hearst_challenge")
  
  puts "Exporting training data."
  o = ExportTrain.new
  o.load  
  puts "done."
end

File: src/com/cleveralgorithms/hearst/export/ExportTrain.java
package com.cleveralgorithms.hearst.export;

import java.io.FileWriter;
import java.io.IOException;
import java.sql.Connection;
import java.sql.DatabaseMetaData;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;




/**
 * Export training stores
 * @author jasonb
 *
 */
public class ExportTrain {
	
	public final static String DB_CONNECTION_STRING = "jdbc:mysql://localhost/hearst_challenge";
	public final static String DB_DRIVER = "com.mysql.jdbc.Driver";
	public final static String DB_USERNAME = "root";
	public final static String DB_PASSWORD = "";
	
	public final static String EXPORT_FILENAME = "experiments/export/export_train.dat";
	
	protected FileWriter writer;
	
	public void run() 
	{
		Connection conn = null;
		try {
			conn = getConnection();
			openFile();
			
			Statement st = conn.createStatement();
			st.setFetchSize(1000);
		    ResultSet rs = st.executeQuery("select store_key,title_key,on_year,on_month,sales_total from template_mo");
		    int count=0;
		    StringBuilder buf = new StringBuilder();
		    while (rs.next()) 
		    {
		    	buf.append(rs.getInt("store_key"));
		    	buf.append(",");
		    	buf.append(rs.getInt("title_key"));
		    	buf.append(",");
		    	buf.append(rs.getInt("on_year"));
		    	buf.append(",");
		    	buf.append(rs.getInt("on_month"));
		    	buf.append(",");
		    	buf.append(rs.getDouble("sales_total"));
		    	buf.append("\n");
		    	
		    	count++;
		    	if ((count%20000)==0) {
		    		writeTextFile(buf.toString());
		    		buf = new StringBuilder(); 
		    		System.out.println(" "+count+" records");
		    	}
		    }
			
		    System.out.println("Wrote data to " + EXPORT_FILENAME);
		    System.out.println("done.");
			
		} catch(Exception e) {
			throw new RuntimeException("Fatal Exception: "+ e.getMessage(), e);
		} finally {
			try {conn.close();} catch(Exception e) {}
			try {closeFile();} catch(Exception e) {}
		}
	}
	
	
	
	protected void openFile() throws IOException
	{
		writer = new FileWriter(EXPORT_FILENAME);
	}
	
	protected void writeTextFile(String s) throws IOException
	{
		writer.write(s);
		writer.flush();
	}
	
	protected void closeFile() throws IOException
	{
		writer.close();
	}
	
	public Connection getConnection() throws SQLException, IllegalAccessException, InstantiationException, ClassNotFoundException
	{
		Class.forName(DB_DRIVER).newInstance();
	    Connection conn = DriverManager.getConnection(DB_CONNECTION_STRING, DB_USERNAME, DB_PASSWORD);
	    
	    return conn;
	}
	
	public static void main(String[] args) {
		new ExportTrain().run();
	}
	
}
